### DISCUSSION TITLE
Difference between ValidationProblem method in Results and TypedResults classes

### DISCUSSION THREAD

**Discussion by McMlok** [dotnet/aspnetcore#54553] [Q&A] [answered]
Hello,

I want to know why methods ValidationProblem  produces different output when they are called on Results or TypedResults class.

Results.ValidationProblem(errorMessages) produces {Microsoft.AspNetCore.Http.HttpResults.ProblemHttpResult} with {Microsoft.AspNetCore.Http.HttpValidationProblemDetails} in detail

TypedResults.ValidationProblem(errorMessages) produces {Microsoft.AspNetCore.Http.HttpResults.ValidationProblem} with {Microsoft.AspNetCore.Http.HttpValidationProblemDetails} in detail.

I have my own IProblemDetailsWriter registered in DI before default one and when I use method on Results class it is hit. But when the TypedResults method is used my writer is not hit. In my code I'm using Results but I want to use [MinimalApis.Extensions](https://github.com/DamianEdwards/MinimalApis.Extensions) for input parameters validation with DataAnnotations. And there is used TypedResults.

```csharp
using Microsoft.Extensions.DependencyInjection.Extensions;
using System.Net;

var builder = WebApplication.CreateBuilder(args);

builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IProblemDetailsWriter, ValidationDetailsWriter>());
builder.Services.AddProblemDetails();

var app = builder.Build();

app.MapGet("results", () => { return Results.ValidationProblem(new Dictionary<string, string[]> { { "test", ["test"] } }); });
app.MapGet("typedresults", () => { return TypedResults.ValidationProblem(new Dictionary<string, string[]> { { "test", ["test"] } }); });

app.Run();

internal class ValidationDetailsWriter : IProblemDetailsWriter
{
  public bool CanWrite(ProblemDetailsContext context) => true;


  public async ValueTask WriteAsync(ProblemDetailsContext context)
  {
    context.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
    await context.HttpContext.Response.WriteAsync("My response");
  }
}
```

It looks like ProblemHttpResult is somewhere handled with ProblemDetailService but ValidationProblem not. 

Is there a way how to handle ValidationProblem with same ProblemDetailService?

Thank you

**Comment by vernou** [comment, marked as answer]
When the result is of type that implement `IResult`, the response is generated by `IResult.ExecuteAsync`.

It's the case with `ProblemHttpResult` (that is returned by `Results.ValidationProblem`) and `ValidationProblem` (that is returned by `TypedResults.ValidationProblem`).

The method [`ProblemHttpResult.ExecuteAsync`](https://github.com/dotnet/aspnetcore/blob/v8.0.3/src/Http/Http.Results/src/ProblemHttpResult.cs#L49) is :
```csharp
public sealed class ProblemHttpResult : IResult, ...
{
    ...

    public async Task ExecuteAsync(HttpContext httpContext)
    {
        var problemDetailsService = httpContext.RequestServices.GetService<IProblemDetailsService>();
        if (problemDetailsService is null || !await problemDetailsService.TryWriteAsync(new() { HttpContext = httpContext, ProblemDetails = ProblemDetails }))
        {
            await HttpResultsHelper.WriteResultAsJsonAsync(
                httpContext,
                logger,
                value: ProblemDetails,
                ContentType);
        }
    }
}
```

The method [`ValidationProblem.ExecuteAsync`](https://github.com/dotnet/aspnetcore/blob/v8.0.3/src/Http/Http.Results/src/ValidationProblem.cs#L52) is :
```csharp
public sealed class ValidationProblem : IResult, ...
{
    public Task ExecuteAsync(HttpContext httpContext)
    {
        return HttpResultsHelper.WriteResultAsJsonAsync(
                httpContext,
                logger,
                value: ProblemDetails,
                ContentType);
    }
}
```

It's similar, just `ProblemHttpResult` allows you to inject a service to generate a custom response.

---

As hack, it's possible to add a operation filter to convert `ValidationProblem` to `ProblemHttpResult` :
```csharp
app.MapGet("typedresults", () => { return TypedResults.ValidationProblem(new Dictionary<string, string[]> { { "test", ["test"] } }); })
    .AddEndpointFilter(async (efiContext, next) =>
    {
        var result = await next(efiContext);
        if(result is ValidationProblem vp)
        {
            return Results.Problem(vp.ProblemDetails);
        }
        return result;
    });
```

---

Can you specify your need?

**Reply by McMlok** [reply]
Thank you for the hint it works.

What I need? In our company we are using our own schema for ProblemDetail different than RFC. And if I use [MinimalApis.Extensions](https://github.com/DamianEdwards/MinimalApis.Extensions) WithParameterValidation extension method it returns response according to RFC because inside TypedResults.ValidationProblem method is used. I want to be able to use Results classes in an application and if we switch to RFC I only remove my own IProblemDetailsWriter implementation.

**Reply by vernou** [reply]
Indeed, [`WithParameterValidation`](https://github.com/DamianEdwards/MinimalApis.Extensions/blob/v0.11.0/src/MinimalApis.Extensions/Filters/ValidationFilterRouteHandlerBuilderExtensions.cs#L180) return `TypedResults.ValidationProblem` on validation error.

If you just want modify the response body, maybe a custom JSON serializer can do the job.

**Comment by bjornen77** [comment]
There is an open issue and PR that will fix this:
https://github.com/dotnet/aspnetcore/issues/47631
